source/extensions/filters/network/kafka/serialization.h

```diff
@@ -0,0 +1,756 @@
+ * First, the length N is given as an INT32. Then N instances of type T follow.
+ * A null array is represented with a length of -1.
+ */
+
+template <typename RT, typename CT>
+class ArrayBuffer : public Deserializer<NULLABLE_ARRAY<RT>> {
+public:
+  size_t feed(const char*& buffer, uint64_t& remaining) {
+
+    const size_t length_consumed = length_buf_.feed(buffer, remaining);
+    if (!length_buf_.ready()) {
+      // break early: we still need to fill in length buffer
+      return length_consumed;
+    }
+
+    if (!length_consumed_) {
+      required_ = length_buf_.get();
+
+      if (required_ >= 0) {
+        children_ = std::vector<CT>(required_);
+      }
+      if (required_ == NULL_ARRAY_LENGTH) {
+        ready_ = true;
+      }
+      if (required_ < NULL_ARRAY_LENGTH) {
+        throw EnvoyException(fmt::format("invalid array length: {}", required_));
+      }
+
+      length_consumed_ = true;
+    }
+
+    if (ready_) {
+      return length_consumed;
+    }
+
+    size_t child_consumed{0};
+    for (CT& child : children_) {
+      child_consumed += child.feed(buffer, remaining);
+    }
+
+    bool children_ready_ = true;
+    for (CT& child : children_) {
+      children_ready_ &= child.ready();
+    }
+    ready_ = children_ready_;
+
+    return length_consumed + child_consumed;
+  }
+
+  bool ready() const {
+    return ready_;
+  }
+
+  NULLABLE_ARRAY<RT> get() const {
+    if (NULL_ARRAY_LENGTH != required_) {
+      std::vector<RT> result{};
+      result.reserve(children_.size());
+      for (const CT& child : children_) {
+        const RT child_result = child.get();
+        result.push_back(child_result);
+      }
+      return { result };
+    } else {
+      return absl::nullopt;
+    }
+  }
+
+private:
+  constexpr static INT32 NULL_ARRAY_LENGTH{-1};
+
+  Int32Buffer length_buf_;
+  bool length_consumed_{false};
+  INT32 required_;
+  std::vector<CT> children_;
+  bool children_setup_{false};
+  bool ready_{false};
+};
+
+// === NULL BUFFER =============================================================
+
+template <typename RT>
+class NullBuffer : public Deserializer<RT> {
+public:
+
+  size_t feed(const char*&, uint64_t&) {
+    return 0;
+  }
+
+  bool ready() const {
+    return true;
+  }
+
+  RT get() const {
+    return {};
+  }
+
+};
+
+// =============================================================================
+// === SERIALIZER ==============================================================
+// =============================================================================
+
+/**
+ * Encoder serializes provided argument to Kafka format
+ * In case of primitive types, this is done explicitly
+ * In case of composite types (request/response), this is done by calling 'encode' on provided argument
+ */
+
+class Encoder {
+public:
+  template <typename T>
+  size_t encode(const T& arg, char* dst);
```

Comment:
current implementation just assumes there's enough space, what's wrong
what would be the favoured API?
`size_t encode(const T& arg, char* dst, size_t dst_size)` with `dst_size` carrying max size that can be written
or are there any Envoy classes I should utilize instead?